### DISCUSSION TITLE
Seq vs Loki

### DISCUSSION THREAD

**Discussion by clabnet** [datalust/seq-tickets#1828] [General]
Hi, I know Seq from many years and I love it. 
On a new project the IT architect would push the Loki solution, integrated on Grafana dashboard.
What strong points can I propose so that SEQ can be used?
Thank's @nblumhardt

**Comment by nblumhardt** [comment]
Hi Claudio, thanks for dropping us a line!

First, although I've checked in on Loki several times over the past year or two (having been a Grafana fan way back in the day!) I am not an expert on its capabilities. It's without doubt a very useful product in its own right.  My impression is that side by side, Loki and Seq have quite different goals. 

Loki places a lot of emphasis on simplifying operations by indexing only the labels given to log streams, rather than log data itself. I believe that this produces some interesting and valuable optimization opportunities for deployments at massive scales.

By comparison, Seq puts emphasis on being effective for on-the-spot production debugging and analysis. Rather than privilege a single field (labels) in the query language, Seq makes it equally straightforward to navigate logs based on any part of the log payload. When you're chasing a bug or responding to an outage, every bit of tooling friction hurts, and we aim to keep Seq's internal implementation details out of your way as much as possible.

Assuming you're deeply familiar with [how queries are written in Seq](https://docs.datalust.co/docs/the-seq-query-language), one thing to check out might be the [Loki docs for log queries](https://grafana.com/docs/loki/latest/logql/log_queries/). There's no doubt a lot of power there, but also a lot of trivia that necessarily leaks into the user experience. Just as one example, deferring parsing of log lines until query time sounds flexible, but leaves me with questions like - do I need to know whether a log event came in as JSON or some other format in order to start searching for it? If my logs are all structured - do I need to insert `| json` into every query just to parse them? You might find more examples over there that distinguish Loki's aims and values from Seq's.

For me Seq's querying experience is really hard to look past, and makes a big impact on how quickly developers can get up to speed and feel productive getting the job done.

(Interestingly, although we started with usability goals first, Seq still ended up with a very similar low-overhead indexing scheme to Loki's, that also concentrates on narrowing the search space. Seq's signal indexes are essentially "virtual labels": instead of being a fixed set defined at the log source, they're user-defined, on the server side. Signals are much more flexible than labels, while still being extremely cheap to compute and store compared with traditional full-text indexes.)

There's more to Seq than just log search, and more to the Grafana stack than just Loki, but there might not be much point going into all of those details without knowing what's important to you. Let me know if there's anything particular I should dig into!

Also, I'd be happy to jump on a quick call and to demo Seq to your team as well as answer any questions from them - please just ping me via `support@` if you think this would help :-)

Best regards,
Nick

**Comment by clabnet** [comment]
Always precious your responses, 
Thanks all Nick.

Il giorno 24 apr 2023, 10:48, alle ore 10:48, Nicholas Blumhardt ***@***.***> ha scritto:
